superset/utils/oauth2.py

```diff
@@ -0,0 +1,69 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from __future__ import annotations
+
+from datetime import datetime, timedelta
+
+from superset import db
+from superset.db_engine_specs.base import BaseEngineSpec
+
+
+def get_oauth2_access_token(
+    database_id: int,
+    user_id: int,
+    db_engine_spec: type[BaseEngineSpec],
+) -> str | None:
+    """
+    Return a valid OAuth2 access token.
+
+    If the token exists but is expired and a refresh token is available the function will
+    return a fresh token and store it in the database for further requests.
+    """
+    # pylint: disable=import-outside-toplevel
+    from superset.models.core import DatabaseUserOAuth2Tokens
+
+    token = (
+        db.session.query(DatabaseUserOAuth2Tokens)
+        .filter_by(user_id=user_id, database_id=database_id)
+        .one_or_none()
+    )
+    if token is None:
+        return None
+
+    if token.access_token and token.access_token_expiration < datetime.now():
+        return token.access_token
+
+    if token.refresh_token:
+        # refresh access token
+        token_response = db_engine_spec.get_oauth2_fresh_token(token.refresh_token)
+
+        # store new access token; note that the refresh token might be revoked, in which
+        # case there would be no access token in the response
+        if "access_token" in token_response:
+            token.access_token = token_response["access_token"]
+            token.access_token_expiration = datetime.now() + timedelta(
+                seconds=token_response["expires_in"]
+            )
+            db.session.add(token)
```

Comment:
If I'm reading this correctly, this logic is being handled by the webworker, and will cause multiple token refreshes to trigger if multiple queries are executed at the same time. I touch on how we solved this for our own OAuth solution here, where a distributed lock ensures that only a single worker executes the refresh, and other workers wait until that's completed: https://github.com/apache/superset/issues/20300#issuecomment-2018638419

Reply:
Oh, great point, @villebro! But I don't think we have a way of creating a global lock in a default install of Superset, no? Since Redis is not mandatory. We could use the metadata DB for that, but that brings some additional complication (we need to manage timeouts ourselves).

How did you implement it?

Reply:
@betodealmeida I implemented it as a context manager that stores the lock in the key-value table, i.e. in the metadata db. This makes it globally accessible from all workers with minimal overhead. Here's a snippet for how it happens in the code flow ("profile" here denotes the internal OAuth app id):

```python
    try:
        raise_for_invalid_token(tokens=tokens, verify=False)
        return tokens
    except AuthTokenValidationException:
        pass

    try:
        logger.debug(f"Tokens on profile `{profile}` have expired, refreshing")
        with AuthLock(username=username, profile=profile) as lock_id:
            tokens = _do_token_refresh(
                username=username,
                tokens=tokens,
                profile=profile,
            )

            try:
                raise_for_invalid_token(tokens=tokens)
            except TokenValidationException as ex:
                raise TokenActionException(
                    f"Refreshed tokens for profile `{profile}` and user `{username}` are invalid: {ex.message}",
                ) from ex

            logger.debug(
                "Updating tokens on profile `superset` for user `%s` (lock id: %s)",
                username,
                lock_id,
            )
            expiration_dttm = datetime.now() + timedelta(days=TOKEN_EXPIRATION_DAYS)
            UpdateKeyValueCommand(
                resource=JWT_RESOURCE,
                key=get_token_key(
                    resource=JWT_RESOURCE, username=username, profile=profile
                ),
                codec=AuthKeyValueCodec(username),
                value=tokens,
                expires_on=expiration_dttm,
            ).run()
            return tokens
    except CreateAuthLockFailedException:
        logger.debug(
            "Auth lock taken for profile `%s` and user `%s`, retrying fetching of superset tokens",
            profile,
            username,
        )
        raise AuthBackoffException()
```

(I did some redacting here, so there may be typos)

That `AuthLock` tries to create an entry in the key-value table under a specific resource with a specific key, and if it's already there, it raises an exception that we reraise to an exception that backoff can handle. Also, the lock has an expiry, meaning that if anything bad happens, it'll get removed after a short while (we seem to have 20 seconds, which should be ample time to do a refresh round trip).

Reply:
That looks awesome, I'll use it here to refresh the token. Thanks, @villebro!

Reply:
Sounds good - I DM'ed the context manager via Slack so you don't have to reinvent the wheel